vite-plugin-ssr/shared/page-configs/Config.ts

```diff
@@ -39,6 +40,14 @@ type Config = ConfigBuiltIn &
     | VikePackages.ConfigVikeSvelte
   )
 
+// Wrapper that transforms a sync function type into an async one.
+// See https://stackoverflow.com/questions/72916246/how-do-you-create-an-async-version-of-a-typescript-function-type
+type Async<T extends (...args: any) => any> = (...args: Parameters<T>) => Promise<ReturnType<T>>
+
+type OnBeforeRenderSync = <PC extends PageContextBuiltInCommon>(pageContext: PC) => { pageContext: Record<string, any> }
+type OnBeforeRenderAsync = Async<OnBeforeRenderSync>
+type OnBeforeRenderFunction = OnBeforeRenderSync | OnBeforeRenderAsync
```

Comment:
Let's change this into something like

```
type OptionalPromise<T> = T | Promise<T>

type OnBeforeRenderFunction = <PC extends PageContext>(pageContext: PC) => OptionalPromise<{ pageContext: Record<string, any> }>
```

And let's wait for @brillout to have made `PageContext` available as a type here.

We also agreed to touch only this one `react-full-v1` example for now, but to touch other hooks like `onRenderClient` and `onRenderHtml`.